include/xtensor/xcontainer.hpp

```diff
@@ -375,24 +375,36 @@ namespace xt
     }
 
     /**
-     * Transposes the container inplace by permuting the shape with @permute.
-     * @param permute the dimensions with the given permutation
+     * Transposes the container inplace by permuting the shape with @permutation.
+     * @param permutation the dimensions with the given permutation
      */
     template <class D>
-    inline void xcontainer<D>::transpose(const std::vector<size_type>& permute)
+    inline void xcontainer<D>::transpose(const std::vector<size_type>& permutation)
     {
-        if (permute.size() != m_shape.size())
-            throw transpose_error(std::string("Permutation does not have the same size as shape"));
+        if (permutation.size() != m_shape.size())
+        {
+            throw transpose_error("Permutation does not have the same size as shape");
+        }
 
-        // check if axis is twice in permute
-        for (size_type i = 0; i < permute.size(); ++i)
-            for (size_type j = i + 1; j < permute.size(); ++j)
-                if (permute[i] == permute[j])
-                    throw transpose_error(std::string("Permutation contains axis more than once"));
+        // check if axis is twice in permutation
+        for (size_type i = 0; i < permutation.size(); ++i)
+        {
+            for (size_type j = i + 1; j < permutation.size(); ++j)
+            {
+                if (permutation[i] == permutation[j])
+                {
+                    throw transpose_error("Permutation contains axis more than once");
+                }
+            }
+        }
 
-        for (const auto& el : permute)
-            if (el >= dimension() || el < 0)
-                throw transpose_error(std::string("Permutation contains wrong axis"));
+        for (const auto& el : permutation)
+        {
+            if (el >= dimension())
```

Comment:
This check should go in the last for loop, before assigning `temp_strides` and `temp_shape`. Thus you save a traversal of the permutation vector. Since permute contains `size_type` (unsigned integer type), testing the negativity of `e1` is useless.